src/core/unit.c

```diff
@@ -4890,7 +4889,12 @@ int unit_setup_exec_runtime(Unit *u) {
         assert(ec);
 
         /* Try to get it from somebody else */
-        UNIT_FOREACH_DEPENDENCY(other, u, UNIT_ATOM_JOINS_NAMESPACE_OF) {
+        UnitDependencyInfo info;
+        UNIT_FOREACH_DEPENDENCY_WITH_INFO(other, info, u, UNIT_ATOM_JOINS_NAMESPACE_OF) {
+                /* UNIT_JOINS_NAMESPACE_OF is symmetric. */
+                if (info.origin_mask == 0)
+                        continue;
```

Comment:
hmm, i don't understand check? can you elaborate?

Reply:
The comment in the code is updated. PTAL.

Reply:
@poettering 
> but to find a suitable existing runtime object it doesn't really matter in which direction we go.
> 
> i.e. if a has JoinsNamespaceOf=b, and a starts first, it should allocate the namespace first, and b should then use it. And the exact same way the other way.

The current behavior does not work like that. If a.service has JoinsNamespaceOf=b.service, then a.service joins the namespaces of b.service if it is already started. But if a.service is started earlier, then b.service will not join the namespaces of a.service.

> So why have this condition here?

To keep the current behavior explained in the above. Previously, the reference introduced by JoinsNamespaceOf= is one-directional. But with the change, reference become bidirectional. Hence, we need to distinguish which unit introduce the dependency.

> if it is dropped, what changes?

Then, setting JoinsNamespaceOf=b.service in a.service implies JoinsNamespaceOf=a.service for b.service.

I think it is OK to 'fix' the current behavior. WDYT?

Reply:
> The current behavior does not work like that. If a.service has JoinsNamespaceOf=b.service, then a.service joins the namespaces of b.service if it is already started. But if a.service is started earlier, then b.service will not join the namespaces of a.service.

Uh, that's a bug! It was intended to be fully symmetric, i.e. A joins B should be entirely equivalent to B joins A... So if the condition is dropped this is actually a bugfix then.